File: leetcode/LinkedList/leetcode-0203/203.go
package linkedlist

// ListNode Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func removeElements(head *ListNode, val int) *ListNode {
	if head == nil {
		return head
	}

	dummy := &ListNode{0, head}
	pre, loc := dummy, head

	for loc != nil {
		if loc.Val == val {
			pre.Next = loc.Next
			loc = loc.Next
		} else {
			pre = pre.Next
			loc = loc.Next
		}
	}
	return dummy.Next
}

File: leetcode/LinkedList/leetcode-0203/203.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        if (head == null) {
            return head;
        }

        ListNode dummy = new ListNode(0);
        dummy.next = head;

        ListNode pre = dummy, loc = head;
        while (loc != null) {
            if (loc.val == val) {
                pre.next = loc.next;
                loc = loc.next;
            } else {
                pre = pre.next;
                loc = loc.next;
            }
        }
        return dummy.next;
    }
}